### ISSUE TITLE
Timezone - UTC?

### ISSUE THREAD

**Issue by wyattyee** [uncheckederror/Ocean2Ocean#11]
Are you storing/displaying in UTC? I noticed a new entry for 6/18 -- that's tomorrow here on the Left Coast.

**Comment by uncheckederror** [comment]
I'm not sure what timezone the the dates are being stored/retrieved. I will investigate this issue. 
@wyattyee Ideally dates/time would be displayed in the PST timezone, correct?

**Comment by Beej126** [comment]
@uncheckederror some code =)
```cs
//azure app services run in UTC by default, we want to calc dates based in local King County time (Pacific Standard)
var NowInPacificStandardTime = TimeZoneInfo.ConvertTimeFromUtc(DateTime.Now.ToUniversalTime(), TimeZoneInfo.FindSystemTimeZoneById("Pacific Standard Time"));
```
(btw: this is Brent Anderson with KCIT, Beej is nickname =)

**Comment by uncheckederror** [comment]
Hey Brent, thanks for the code sample!

This is issue goes a lot deeper than I expected.

This code worked perfectly on my workstation, a Windows 10 machine, but when I deployed it to the production web server it broke the app.

It turns out that that this is because of a mismatch in the way that my development platform, Windows 10, handles system timezones and the way that the deployment target platform, Azure App Service on Linux, handles them.

This method call "TimeZoneInfo.FindSystemTimeZoneById("Pacific Standard Time")" works on Windows, but returns an exception on Linux. The correct timezone string on Linux is "PST8PDT" which is the version of the PST timezone that respects daylight savings time.

I was able to figure out what was going on here by reading through this thread on setting a time zone on a [Linux Azure App Service instance.](https://github.com/MicrosoftDocs/azure-docs/issues/33417)

Additionally [this blog](https://devblogs.microsoft.com/dotnet/cross-platform-time-zones-with-net-core/) by the ASP.NET Core team was helpful in understanding why the timezone codes are different between Linux and Windows system and how to handle them in this app.

For now I've added an "Application setting" to my Azure App Service instance that sets the systems time to our local time zone.
`Name: TZ Value: PST8PDT`
I have verified that this changed worked by using the web SSH feature to open a terminal on the web service and running the "date" command to get the current system date and time.

A P P   S E R V I C E   O N   L I N U X

Documentation: http://aka.ms/webapp-linux
Dotnet quickstart: https://aka.ms/dotnet-qs
ASP .NETCore Version: 3.1.0
Note: Any data outside '/home' is not persisted
root@3b7b58e0ce59:~/site/wwwroot# date
Fri Jun 19 06:51:23 Pacific 2020

I will have to revisit this issue later, after I figure out how to query the client's timezone in javascript and pass it back to the server, as locking to whole app to the PST timezone seems like a coarse solution to this problem.

**Comment by Beej126** [comment]
shoot, time is always so annoying =)
thanks for sharing the troubleshooting details, very enlightening!
i'm guessing the linux app svc is cheaper/free vs windows?
and totally good point about current time being a client context vs server

**Comment by uncheckederror** [comment]
@Beej126 Yeah, the Linux versions are way cheaper.

Per the [calculator.](https://azure.microsoft.com/en-us/pricing/details/app-service/windows/) The B1 instance I'm using is ~$12.41/month on Linux and ~$54.75/month for the same instance on Windows. 

It's funny though, because the most expensive resource for this app is the Azure SQL instance which is ~$230 a month. So the app is cheap at $10 and the database is 23 times more expensive. 🚀

**Comment by Beej126** [comment]
from this morning's call it sounded like this is getting migrated under KCIT's tab here pretty soon anyway?
if that's still a ways off... i love t-sql too but if you're already down the cheapo linux path what about shifting to mysql or whatever?

**Comment by uncheckederror** [comment]
I talked with Brian Chapman this morning and he showed me how to create the azure resources I need in the sandbox. I will be redeploying this app to the KCIT SE Sandbox tonight. 

I also read some blogs about getting the cheapest Azure SQL instance possible. One the configure page for the SQL database instance there a small option "Looking for basic, standard, and premium?". If you click that, and then select the "Basic" option and set the maximum database size to 100 MB you can get the cost for an Azure SQL database down to $5 a month. At which point the total cost for this app's infrastructure is now ~$15 a month. ✨

Thanks for talking through this with me @Beej126 .

Getting back to the time/date issue I think the longer term solution here is not auto-populate the server-side DateTime on behalf of the user when they submit their steps as the app does now. Rather the date should be a configurable field on the web-page and then we should use javascript to set it by default to the local date of the browser. The advantage of this method is that it no longer matters to the entry what the timezone of the app is, rather the correctness of the date is now a client-side rather than server-side concern. This would also enable people to submit their steps for prior days which I think could be useful.

**Comment by uncheckederror** [comment]
This app is now running under the KC Sandbox Azure subscription and I have resolved the datetime issue in #14 .
